### PR TITLE
Fix dashboard data processing

### LINKED ISSUE #423: Errors in data manipulation of the `carbonboard`
* CodeCarbon version: 2.2.4
* Python version: Python 3.10.12
* Operating System: MacOS 13.4.1 

### Description

I just discovered this very cool project, thank you for creating it! 🙂

I am currently testing `codecarbon` locally to see how it works, and I have noticed some possible bugs with the data manipulation in the local `carbonboard`.

### What I Did

- Created a very simple script that uses `EmissionsTracker` to track energy usage.
- Run `carbonboard --filepath="emissions.csv" --port=3333` to visualise the results

### Errors

These are some of the Exceptions that I saw in the Dash app logs:

```
  File "/[path]/miniconda3/envs/code-carbon-test/lib/python3.10/site-packages/codecarbon/core/emissions.py", line 228, in _global_energy_mix_to_emissions_rate
    energy_sum = energy_mix["total_TWh"]
KeyError: 'total_TWh'
```

```
  File "/[path]/miniconda3/envs/code-carbon-test/lib/python3.10/site-packages/codecarbon/viz/data.py", line 130, in get_global_emissions_choropleth_data
    global_energy_mix[country_iso_code]["fossil_TWh"], total
KeyError: 'fossil_TWh'
```

```
  File "/[path]/miniconda3/envs/code-carbon-test/lib/python3.10/site-packages/pandas/core/methods/to_dict.py", line 207, in to_dict
    raise ValueError(f"orient '{orient}' not understood")
ValueError: orient 'rows' not understood
```

### (Possible) Solutions

I have noticed that the keys/column names of the data used in the dashboard have changed in https://github.com/mlco2/codecarbon/pull/350/files#diff-729fd2eaf00d62e16106163521e91a169e8a24cf9d17250b5ff6bfc6cabafcda

The names affected should be updated and also the `orient="rows"` should be changed.

I will make a PR to solve this if you agree :)

### PR DESCRIPTION
**Pull request by EdAbati**
Fixes #423 

Solutions:
- Fixed `pandas.DataFrame.to_dict("records")`
- Updated column names to reflect data in https://github.com/mlco2/codecarbon/blob/master/codecarbon/data/private_infra/global_energy_mix.json
- Added unit tests to catch these tyupe of issues in the future
- Extracted some manipulation in `get_country_choropleth_data` to make it easier to test

Please let me know if this is something you are interested in, and if I can improve this PR in any way :)

### CODE REVIEW

**Comment by benoit-cty** [comment]
Thanks, while testing your PR, I find other bugs :
- per_capita_electricity is not in TWh, it is in Wh
- per_capita_electricity is used in total_TWh computation, so it get it completely wrong, so the percentage was wrong in the end.
- total_TWh also use *_Twh, but there are already other aggregates.
- Errors in `emissions.csv`  : azure and aws do not provide carbon intensity of their cloud.

I've pushed all these fixes, could you test it on your side ?

File: codecarbon/data/private_infra/our_world_in_data-2021_data.ipynb
Comment (EdAbati) [inline codecarbon/data/private_infra/our_world_in_data-2021_data.ipynb:1]: I think that the comment at the end of the notebook _"The output we need is a JSON file named `global_energy_mix.json` in this format:"_ is outdated now. The data has a different structure.

File: codecarbon/data/private_infra/global_energy_mix.json
Comment (EdAbati) [inline codecarbon/data/private_infra/global_energy_mix.json:1]: For consistency, we could translate `"gaz_TWh"` to English (`gas_TWh"`)

**Review by EdAbati — COMMENTED** [review]
Hi @benoit-cty, thank you for the rest of the changes, I have tested the dashboard and it works :)

I have noticed some other potential problems (please see my comments). 
Let me know what do you think. I am happy to make the fixes in the PR or in separate PRs, depends of what you prefer.

File: examples/emissions.csv [hunk 1]
```diff
@@ -78,24 +78,24 @@ timestamp,project_name,duration,emissions,emissions_rate,cpu_power,gpu_power,ram
...
 2020-07-19T03:26:11,project_delta,38,9,4.5,2,0,1.234538259,2,0,0.156533891,2,Taiwan,TWN,Changhua County,Y,gcp,asia-east1,,,,,,,,,,,,,,
-2020-07-20T09:26:11,project_gamma,45,4,2,9,0,1.234538293,9,0,0.156533898,9,USA,USA,Ohio,Y,aws,us-east-2,,,,,,,,,,,,,,
+2020-07-20T09:26:11,project_gamma,45,4,2,9,0,1.234538293,9,0,0.156533898,9,USA,USA,Ohio,N,,,,,,,,,,,,,,,,
 2020-07-21T15:26:11,project_beta,14,1,0.5,8,0,1.234538327,8,0,0.156533905,8,Canada,CAN,Ontario,N,,,,,,,,,,,,,,,,
 2020-07-22T21:26:11,project_alpha,48,4,2,6,0,1.234538361,6,0,0.156533913,6,Canada,CAN,Ontario,N,,,,,,,,,,,,,,,,
-2020-07-24T03:26:11,project_theta,32,8,4,5,0,1.234538396,5,0,0.15653392,5,Brazil,BRA,Sao Paulo,Y,azure,brazilsouth2021-07-14T10:59:36,codecarbon,3565.51492,0.027399096,2.6775,0,1.533303325,0.022284099,0,0.012761264,0.035045363,Morocco,MAR,casablanca-settat,N
\ No newline at end of file
+2020-07-24T03:26:11,project_theta,32,8,4,5,0,1.234538396,5,0,0.15653392,5,Brazil,BRA,Sao Paulo,N,,brazilsouth2021-07-14T10:59:36,codecarbon,3565.51492,0.027399096,2.6775,0,1.533303325,0.022284099,0,0.012761264,0.035045363,Morocco,MAR,casablanca-settat,N
```
[the remark this answers is not in the document]
Comment (benoit-cty) [reply examples/emissions.csv:101]: Good point, I think the extra data needs to be deleted as they don't refer to an example name.

File: tests/test_viz_data.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        "iso_code": "TST",
+        "low_carbon_TWh": 110.0,
+        "nuclear_TWh": 20.0,
+        "oil_TWh": 15.0,
+        "other_renewable_TWh": 25.0,
+        "other_renewable_exc_biofuel_TWh": 5.0,
+        "per_capita_TWh": 500.0,
+        "renewables_TWh": 110.0,
+        "solar_TWh": 20.0,
+        "total_TWh": 1200.0,
```
[the remark this answers is not in the document]
Comment (benoit-cty) [reply tests/test_viz_data.py:75]: Yes, exactly, _total_TWh_ is totally wrong due to _per_capita_TWh_.

File: codecarbon/data/private_infra/global_energy_mix.json
Comment (benoit-cty) [reply codecarbon/data/private_infra/global_energy_mix.json:1]: Yes, sorry for that, here in France we think that words ending with Z is English  :smile:

File: codecarbon/data/private_infra/our_world_in_data-2021_data.ipynb
Comment (benoit-cty) [reply codecarbon/data/private_infra/our_world_in_data-2021_data.ipynb:1]: Yes, maybe it's better to delete it : the actual JSON will show the expected format.

File: setup.py [hunk 3]
```diff
@@ -12,14 +12,15 @@
...
+    "prometheus_client",
 ]
 
 TEST_DEPENDENCIES = ["mock", "pytest", "responses", "tox", "numpy", "requests-mock"]
 
 
 setuptools.setup(
     name="codecarbon",
-    version="2.2.4",
+    version="2.2.5",
```
[the remark this answers is not in the document]
Comment (benoit-cty) [reply setup.py:23]: Good point, it's better with `">=3.7"`

**Comment by benoit-cty** [comment]
> Let me know what do you think. I am happy to make the fixes in the PR or in separate PRs, depends of what you prefer.

You could do everything in this PR. Thanks a lot !

**Comment by EdAbati** [comment]
Hi @benoit-cty , sorry for the delay! This should be ready now. Let me know if there is somenthing else I should change :)

**Review by benoit-cty — APPROVED** [review]
Thanks a lot, I will publish a new release with you PR.

**Pull request merged**